### PR TITLE
Lint Readable 

### PR DESCRIPTION
**Pull request by GalRabin**
@anara123 
@yaakovi 
@yuvalbenshalom 
@bakatzir 

## Status
In Progress

## Related Issues
fixes: https://github.com/demisto/etc/issues/21345
fixes: https://github.com/demisto/etc/issues/22148
fixes: https://github.com/demisto/etc/issues/21750 @bakatzir 
fixes: https://github.com/demisto/etc/issues/21569
reference: https://github.com/demisto/etc/issues/22759

## Description
Lint command readable output.
Until today the lint only printed the raw output to the screen, due to that an overall optimization required in order to parse the data and supply a parsed output.

### Changes:
1. Logging option -v ... -vvvvvv (verbosity levels - logging levels verbose-critical- .. - debug)
2. Logging path option - Until debug level.
3. Docker-sdk, git-python usage for better exceptions handling and output,
4. overall optimizations and code readability.
5. JSON output including all collected data for analytics.
6. Adding the following flags:
* test-xml - export test results as xml.
* json-report - export json report on all results.
* log - save the log to file.
7. Git python pack filtering - using set operations (save 19 sec- - each time).
8. Facts gathering optimization - Pipfile collecting, modules from master, additional packages adding to each package 'test-requirements.txt` in the package.
9. Context manager for mandatory test modules (CommonServerPython etc).
10. Code readability.
11. Docker image builds and not commit (Not the recommended way), Now we are using only Single Docker file to the setup test image.
12. Adding many unit-tests in order to be more resilient, exceptions, fixtures etc - https://github.com/demisto/demisto-sdk/tree/lint/demisto_sdk/commands/lint/tests.

## CI example
- When running against branch (no errors) - https://circleci.com/gh/demisto/content/44338#queue-placeholder/containers/1
- When running against branch (pytest error trigered) - 
https://circleci.com/gh/demisto/content/44341#queue-placeholder/containers/1
- When running on all packages (pytest error trigered)  - https://app.circleci.com/pipelines/github/demisto/content/4617/workflows/5cb54459-d3b2-482f-b1ac-7f345d2d1676/jobs/44355 
** We need to provide solution for vulture errors.
**Log example** https://44355-60525392-gh.circle-artifacts.com/1/artifacts/lint_debug_log.log (If exception occured-search for critical, error - if lint occurred"
**json report example**- https://44355-60525392-gh.circle-artifacts.com/1/artifacts/lint_report.json (Sort - by pack name)
**Test failure presentation of CircleCi**- https://app.circleci.com/pipelines/github/demisto/content/4617/workflows/5cb54459-d3b2-482f-b1ac-7f345d2d1676/jobs/44355/tests

### Change that should be done in config.yml in content repo:
https://github.com/demisto/content/pull/6099

### CODE REVIEW

**Comment by fvigo** [comment]
During the refactoring, please make sure that the error checks are in place. For example, running `demisto-sdk lint` on a machine that doesn't have docker returns this error, which is not very clear:

```
Traceback (most recent call last):
  File "/usr/local/bin/demisto-sdk", line 11, in <module>
    load_entry_point('demisto-sdk==0.4.2', 'console_scripts', 'demisto-sdk')()
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/demisto_sdk-0.4.2-py3.7.egg/demisto_sdk/__main__.py", line 321, in lint
    return linter.run_dev_packages()
  File "/usr/local/lib/python3.7/site-packages/demisto_sdk-0.4.2-py3.7.egg/demisto_sdk/commands/lint/lint_manager.py", line 126, in run_dev_packages
    run_status_code = linter.run_dev_packages()
  File "/usr/local/lib/python3.7/site-packages/demisto_sdk-0.4.2-py3.7.egg/demisto_sdk/commands/lint/linter.py", line 134, in run_dev_packages
    py_num = get_python_version(dockers[0], self.log_verbose)
  File "/usr/local/lib/python3.7/site-packages/demisto_sdk-0.4.2-py3.7.egg/demisto_sdk/commands/common/tools.py", line 453, in get_python_version
    universal_newlines=True, stderr=stderr_out).strip()
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 411, in check_output
    **kwargs).stdout
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['docker', 'run', '--rm', 'demisto/python3:3.7.4.2245', 'python', '-c', "import sys;print('{}.{}'.format(sys.version_info[0], sys.version_info[1]))"]' returned non-zero exit status 125.
```

**Comment by GalRabin** [comment]
> he refactoring, please make sure that the error checks are in place. For example, running `demisto-sdk lint` on a machine that doesn't have docker returns this error, which is not ve
Fixed in the refactor

**Comment by anara123** [comment]
@glicht FYI

File: demisto_sdk/__main__.py [hunk 1]
```diff
@@ -266,52 +266,67 @@ def secrets(config, **kwargs):
...
+@click.option("-p", "--parallel", default=1, help="Run tests in parallel")
+@click.option("--no-flake8", is_flag=True, help="Do NOT run flake8 linter")
+@click.option("--no-bandit", is_flag=True, help="Do NOT run bandit linter")
+@click.option("--no-mypy", is_flag=True, help="Do NOT run mypy static type checking")
+@click.option("--no-vulture", is_flag=True, help="Do NOT run mypy static type checking")
+@click.option("--no-pylint", is_flag=True, help="Do NOT run pylint linter")
+@click.option("--no-test", is_flag=True, help="Do NOT test (skip pytest)")
+@click.option("-kc", "--keep-container", is_flag=True, help="Keep the test container")
+@click.option("--test-xml", help="Path to store pytest xml results", type=click.Path(exists=True))
+@click.option("--json-report", help="Path to store pytest xml results", type=click.Path(exists=True))
```
[the remark this answers is not in the document]
Comment (GalRabin) [reply demisto_sdk/__main__.py]: Yes, will be fixed

**Comment by glicht** [comment]
Looking at the all tests link you provided. Run time has gone up to 1:28. Nightly build runs all tests in 43 minutes.

**Comment by glicht** [comment]
When ready for review please provide a link to a run with all tests (nightly style run)

**Comment by GalRabin** [comment]
@bakatzir @glicht @anara123 @yuvalbenshalom @yaakovi @ShahafBenYakir 
Update:
1. Powershell test and analyze support.
2. Debug logging (-vvvvvv) will print output of commands (even if not failing), the log exported in Circle CI will always output debug.
3. Docker push to docker hub after building - thanks to @glicht 

Circle CI examples:
1. Comparing to branch - 6 packs changed:
     a. Steps (1:33 min) - https://circleci.com/gh/demisto/content/44925#queue-placeholder/containers/1
     b. logs - https://44925-60525392-gh.circle-artifacts.com/1/artifacts/lint_debug_log.log
     c. report - https://44925-60525392-gh.circle-artifacts.com/1/artifacts/lint_report.json
     d. tests page - Press on `Tests` tab.
2. Nightly run - all packages run:
     a. Steps (43:51 min) - https://circleci.com/gh/demisto/content/44910#queue-placeholder/containers/1
     b. logs - https://44910-60525392-gh.circle-artifacts.com/1/artifacts/lint_debug_log.log
     c. report - https://44910-60525392-gh.circle-artifacts.com/1/artifacts/lint_report.json
     d. tests page - Press on `Tests` tab.

File: demisto_sdk/commands/lint/helpers.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+            try:
+                container_obj.remove(force=True)
+                break
+            except docker.errors.APIError:
+                continue
+
+    return py_num
+
+
+def run_command_os(command: str, cwd: Path) -> Tuple[str, str, int]:
```
Comment (bakatzir) [inline demisto_sdk/commands/lint/helpers.py]: please delete and use the version we already have in `tools.py`

File: demisto_sdk/commands/lint/helpers.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    "bandit": 0b10,
+    "mypy": 0b100,
+    "vulture": 0b1000000,
+    "pytest": 0b1000,
+    "pylint": 0b10000,
+    "image": 0b100000,
+    "pwsh_analyze": 0b10000000,
+    "pwsh_test": 0b100000000
+}
+
```
Comment (bakatzir) [inline demisto_sdk/commands/lint/helpers.py]: please add consts for skip and exit code - 0b0, 0b10

File: demisto_sdk/commands/lint/linter.py [hunk 4]
```diff
@@ -1,533 +1,803 @@
...
-        if self.script_type == TYPE_PYTHON:
-            md5 = hashlib.md5(requirements.encode('utf-8') + setup_script_data).hexdigest()
+        log_prompt = f"{self._pack_name} - Image create"
+        test_image_id = ""
+        # Get requirements file for image
+        if 2 < docker_base_image[1] < 3:
+            requirements = self._req_2
+        elif docker_base_image[1] > 3:
+            requirements = self._req_3
         else:
```
Comment (bakatzir) [inline demisto_sdk/commands/lint/linter.py]: how will we reach here?

File: demisto_sdk/__main__.py [hunk 5]
```diff
@@ -258,52 +258,75 @@ def secrets(config, **kwargs):
...
-    return linter.run_dev_packages()
+              short_help="Run lintings (flake8, mypy, pylint, bandit) and pytest. pylint and pytest will run within the"
+                         "docker image of an integration/script. Meant to be used with integrations/scripts that use "
+                         "the folder (package) structure. Will lookup up what docker image to use and will setup the "
+                         "dev dependencies and file in the target folder. ")
+@click.help_option('-h', '--help')
+@click.option("-d", "--dir-pack", help="Specify directory of integration/script", type=click.Path(exists=True,
+                                                                                                  resolve_path=True))
+@click.option("-g", "--git", is_flag=True, help="Will run only on changed packages")
+@click.option("-a", "--all-packs", is_flag=True, help="Run lint on all directories in content repo")
```
[the remark this answers is not in the document]
Comment (bakatzir) [reply demisto_sdk/__main__.py]: same behaviour as in validate

[the remark this answers is not in the document]
Comment (glicht) [reply demisto_sdk/__main__.py]: Would rather not. Running by default on all is a mess. Killing this doesn't always even work as processes and dockers are spawned. It should be explicitly stated the  `-a` to run on all.

File: demisto_sdk/commands/lint/lint_manager.py [hunk 6]
```diff
@@ -1,301 +1,504 @@
...
+        # Validating docker engine connection
+        docker_client: docker.DockerClient = docker.from_env()
+        try:
+            docker_client.ping()
+        except (requests.exceptions.ConnectionError, urllib3.exceptions.ProtocolError, docker.errors.APIError):
+            facts["docker_engine"] = False
+            print_error("Can't communicate with Docker daemon - check your docker Engine is ON - Skiping lint, "
+                        "test which require docker!")
+            logger.info(f"demisto-sdk-lint - Can't communicate with Docker daemon")
+        if not test_internet_connection():
```
[the remark this answers is not in the document]
Comment (glicht) [reply demisto_sdk/commands/lint/lint_manager.py]: This check is wrong. I can run docker tests without an internet connection if I have the image already setup on my machine. Would rather we don't do this check and fail on the actual error encountered.

File: demisto_sdk/commands/lint/helpers.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    modules = [Path("Tests/demistomock/demistomock.py"),
+               Path("Tests/scripts/dev_envs/pytest/conftest.py"),
+               Path("Packs/Base/Scripts/CommonServerPython/CommonServerPython.py"),
+               Path("Tests/demistomock/demistomock.ps1"),
+               Path("Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1")]
+    modules_content = {}
+    if content_repo:
+        remote = content_repo.remote()
+        for module in modules:
+            modules_content[module] = content_repo.commit(f'{remote.name}/master').tree[
```
Comment (glicht) [inline demisto_sdk/commands/lint/helpers.py]: What are you doing here? Please provide an explanation.

File: demisto_sdk/commands/lint/lint_manager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (GalRabin) [reply demisto_sdk/commands/lint/lint_manager.py]: @glicht , I agree with it, I will remove this test and I will handle exceptions if occured due to no internet connection.

File: demisto_sdk/commands/lint/lint_manager.py [hunk 8]
```diff
@@ -1,301 +1,504 @@
...
+            facts["docker_engine"] = False
+            print_error("No internet connection - Skiping lint, test which require docker!")
```
[the remark this answers is not in the document]
Comment (GalRabin) [reply demisto_sdk/commands/lint/lint_manager.py]: Remove - so not relevant.

File: demisto_sdk/commands/lint/lint_manager.py [hunk 9]
```diff
@@ -1,301 +1,504 @@
...
+        logger.info(f"lint - Docker daemon test passed")
+
+        return facts
+
+    def _get_packages(self, content_repo: git.Repo, dir_packs: str, git: bool, all_packs: bool) -> List[Path]:
```
[the remark this answers is not in the document]
Comment (GalRabin) [reply demisto_sdk/commands/lint/lint_manager.py]: No, I am using varbose there.

File: demisto_sdk/commands/lint/helpers.py [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+                str(module)].data_stream.read()
+    else:
+        for module in modules:
+            url = f'https://raw.githubusercontent.com/demisto/content/master/{module}'
+            modules_content[module] = requests.get(url=url,
```
Comment (glicht) [inline demisto_sdk/commands/lint/helpers.py]: doesn't look like you are checking for requests status code. For example you could get a 404 and not notice.

File: requirements.txt [hunk 11]
```diff
@@ -3,14 +3,18 @@ bs4>=0.0.1
 click>=7.0
 dateparser>=0.7.2
 demisto-py>=2.0.7
+docker==4.2.0
```
Comment (glicht) [inline requirements.txt]: probably better to use '>='

File: requirements.txt [hunk 12]
```diff
@@ -3,14 +3,18 @@ bs4>=0.0.1
...
 flake8>=3.7.9
-gitpython>=3.1.0
+gitpython>=3.1.
```
Comment (glicht) [inline requirements.txt]: Why did you remove the 0?

File: requirements.txt [hunk 13]
```diff
@@ -3,14 +3,18 @@ bs4>=0.0.1
...
+jinja2==2.11.1
```
Comment (glicht) [inline requirements.txt]: Probably better to do >=

File: demisto_sdk/commands/lint/helpers.py (hunk 7 above)
Comment (GalRabin) [reply demisto_sdk/commands/lint/helpers.py]: I will add an explaination, I am trying to get the file from the local copy of the remote repo <remote_name>/master before downloading directly from GitHub, So it will be possible to work without internet connection.

File: demisto_sdk/commands/lint/helpers.py (hunk 10 above)
Comment (GalRabin) [reply demisto_sdk/commands/lint/helpers.py]: Change it to be handled with res.ok

Comment (GalRabin) [reply demisto_sdk/commands/lint/helpers.py]: Also add trials, so it will not fail on first trial

File: demisto_sdk/commands/lint/commands_builder.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        https://flake8.pycqa.org/en/latest/user/invocation.html
+    Args:
+        files(List[Path]): files to execute lint
+
+    Returns:
+        str: flake8 command
+    """
+    max_line_len = 130
+    # Ignoring flake specific errors https://flake8.pycqa.org/en/latest/user/error-codes.html
+    errors_ignoring = ["W293", "W504", "W291", "W605", "F405", "F403", "E999", "W503", "F841", "E302", "C901", "F821",
```
Comment (glicht) [inline demisto_sdk/commands/lint/commands_builder.py]: Why is this here? We control flake8 ignores and line length in this file: https://github.com/demisto/content/blob/master/tox.ini

Comment (GalRabin) [reply demisto_sdk/commands/lint/commands_builder.py]: Didn't noticed that, Changed

**Comment by glicht** [comment]
See my comments here: https://github.com/demisto/etc/issues/21345#issuecomment-605641740

File: demisto_sdk/commands/lint/helpers.py (hunk 7 above)
Comment (glicht) [reply demisto_sdk/commands/lint/helpers.py]: We should just use the file on disk not from `master`. Otherwise it is impossible to test a change to one of these files.

File: demisto_sdk/commands/common/dev_sh_scripts/pkg_dev_container_setup_pwsh.sh [hunk 15]
```diff
@@ -1,18 +0,0 @@
-#!/bin/sh
-
-# Setup a container for dev testing for pwsh
```
Comment (glicht) [inline demisto_sdk/commands/common/dev_sh_scripts/pkg_dev_container_setup_pwsh.sh]: What code replaced this file?

Comment (GalRabin) [reply demisto_sdk/commands/common/dev_sh_scripts/pkg_dev_container_setup_pwsh.sh]: Dockerfile template - https://github.com/demisto/demisto-sdk/blob/lint/demisto_sdk/commands/lint/templates/dockerfile.jinja2#L29

Build executed here - https://github.com/demisto/demisto-sdk/blob/lint/demisto_sdk/commands/lint/linter.py#L490

File: demisto_sdk/commands/lint/helpers.py (hunk 7 above)
Comment (GalRabin) [reply demisto_sdk/commands/lint/helpers.py]: I am not chaning the file it exists in the directory, just if it missing, If it exists I skip creating the file there, I handle it, in  the following context manager - https://github.com/demisto/demisto-sdk/blob/lint/demisto_sdk/commands/lint/helpers.py#L164

Currently every time when using the lint it create the file from scratch (demisto-mock.py which is not wanted behavior)

**Comment by GalRabin** [comment]
> See my comments here: [demisto/etc#21345 (comment)](https://github.com/demisto/etc/issues/21345#issuecomment-605641740)

I comment also the issue, I fixed the issues

File: demisto_sdk/commands/lint/helpers.py (hunk 7 above)
Comment (GalRabin) [reply demisto_sdk/commands/lint/helpers.py]: @glicht 
Get it from disk

**Comment by glicht** [comment]
@GalRabin build is failing

**Comment by GalRabin** [comment]
> @GalRabin build is failing

Its my unit-tests, After I will finish getting bugs from the developers I will make them work correctly, Thanks

**Comment by GalRabin** [comment]
Updating on changes done until now:
1. Reducing the size of docker images - Reduce the running time to 37:00 min nightly instead of 43:00 min
2. Verbosity level by default on -vv (by @glicht request)
3. Better user experience - tested on daily job of 12 developers, Recieved many feedbacks from the users regarding the output and I changed it to satisfied their needs.
4. Bugs found in  -
*  Powershell analyze and test has been fixed.
*  Git filters has been optimized and tested and diffrent enviorments.

**Comment by GalRabin** [comment]
Nightly circle ci tests (The optimal number of threads to be in use is 8, From tests I perfromed, Due to GIL):
* 8 Threads  (37:06 min) - https://circleci.com/gh/demisto/content/46415#tests/containers/1
* 8 Threads (36:46 min) - https://circleci.com/gh/demisto/content/46344#tests/containers/1

File: demisto_sdk/__main__.py [hunk 16]
```diff
@@ -268,52 +268,58 @@ def secrets(config, **kwargs):
...
+@click.option('-v', "--verbose", count=True, help="Verbosity level -v / -vv / .. / -vvv",
+              type=click.IntRange(0, 3, clamp=True), default=2, show_default=True)
+@click.option('-q', "--quiet", is_flag=True, help="Quiet output, only output results in the end")
+@click.option("-p", "--parallel", default=1, help="Run tests in parallel", type=click.IntRange(0, 15, clamp=True),
+              show_default=True)
+@click.option("--no-flake8", is_flag=True, help="Do NOT run flake8 linter")
+@click.option("--no-bandit", is_flag=True, help="Do NOT run bandit linter")
+@click.option("--no-mypy", is_flag=True, help="Do NOT run mypy static type checking")
+@click.option("--no-vulture", is_flag=True, help="Do NOT run mypy static type checking")
```
Comment (yaakovi) [inline demisto_sdk/__main__.py:288]:
```suggestion
@click.option("--no-vulture", is_flag=True, help="Do NOT run vulture linter")
```
